### ISSUE TITLE
Emojis are not always displayed in the default font

### ISSUE THREAD

**Issue by papojari** [alacritty/alacritty#5589]
Even though I have defined Noto Color Emoji as the font for emojis alacritty doesn't always display emojis in that font.

![image](https://user-images.githubusercontent.com/81317317/140164252-92fa1e11-a7e7-4ab0-88c4-06ed1ff52100.png)

### System

OS: Linux
Version: 0.9.0
Linux/BSD: Wayland, Sway

### Logs

Font size: 15

**Comment by chrisduerr** [comment]
If the font doesn't have the glyphs, it will use fallback fonts instead. That's to be expected.

**Comment by maxreuben** [comment]
Is there anyway to change the fallback font?

**Comment by chrisduerr** [comment]
On Linux it's Fontconfig.

**Comment by papojari** [comment]
I have set Iosevka as the monospace font and Noto Color Emoji as the emoji font. In kitty both of those fonts are used. In alacritty some emojis are displaye in Noto Color Emoji and some not as seen in the picture above.

**Comment by papojari** [comment]
![image](https://user-images.githubusercontent.com/81317317/148693336-269fbdfa-f537-46ad-858b-251e7a6b6107.png)

**Comment by maxreuben** [comment]
I did a little bit of googling and [here](https://askubuntu.com/questions/1029661/18-04-color-emoji-not-showing-up-at-all-in-chrome-only-partially-in-firefox/1029675#1029675)'s what worked for me. 
Now that I think of it, it was easy but I didn't know about fontconfig before.

**Comment by papojari** [comment]
I already have my font config set up with an Emoji font. The problem here must be Alacritty because in Kitty the fonts all work great without any extra config.

**Comment by maxbrunet** [comment]
I have found the issue for myself, Alacritty does not use the configured `emoji` font family, so a configuration like this one has no effect:

```xml
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'urn:fontconfig:fonts.dtd'>
<fontconfig>

  <alias>
    <family>sans-serif</family>
    <prefer>
      <family>Noto Sans</family>
    </prefer>
  </alias>

  <alias>
    <family>serif</family>
    <prefer>
      <family>Noto Serif</family>
    </prefer>
  </alias>

  <alias>
    <family>monospace</family>
    <prefer>
      <family>Source Code Pro</family>
    </prefer>
  </alias>

  <alias>
    <family>emoji</family>
    <prefer>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
</fontconfig>
```

The fallback to the emoji font needs to be explicit, this works for example:

```xml
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'urn:fontconfig:fonts.dtd'>
<fontconfig>

  <alias>
    <family>sans-serif</family>
    <prefer>
      <family>Noto Sans</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>

  <alias>
    <family>serif</family>
    <prefer>
      <family>Noto Serif</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>

  <alias>
    <family>monospace</family>
    <prefer>
      <family>Source Code Pro</family>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>

  <alias>
    <family>emoji</family>
    <prefer>
      <family>Noto Color Emoji</family>
    </prefer>
  </alias>
</fontconfig>
```

Alacritty will actually fall through the `sans-serif` family (one may have expected it to be `monospace` instead?). I wonder if there is anything to improve here

**Comment by chrisduerr** [comment]
> Alacritty will actually fall through the sans-serif family (one may have expected it to be monospace instead?). I wonder if there is anything to improve here

It should be `fc-match -s "monospace"` by default. So fall through the `monospace` by default. And by should I mean it does on my system at least.

**Comment by papojari** [comment]
Why does it fall through monospace and not just use the emoji font for all emojis in the first place? Maybe Iosevka (my monospace font) has emojis for some characters which then are what annoys me and the rest is displayed in Noto Color Emoji.

**Comment by chrisduerr** [comment]
Because the user has selected `monospace` as their font? There's no such thing as a universal emoji font identifier and many glyphs might be present as emojis in many fonts.

**Comment by maxbrunet** [comment]
Interesting, `fc-match -s "monospace"` does return the `monospace` fonts, but Alacritty does `fc-match -s "family=monospace:style=Regular"` which returns the `sans-serif` for me :confused:

https://github.com/alacritty/crossfont/blob/7db4032f7150a8abe4893b85b97e07715636271a/src/ft/fc/mod.rs#L280-L281

Maybe it should be `fc-match -s "monospace:style=Regular"` which works fine. Note the `family=` has been removed, the `man` page example does not have it: https://www.mankier.com/1/fc-match#Examples

This line could be the culprit: https://github.com/alacritty/crossfont/blob/7db4032f7150a8abe4893b85b97e07715636271a/src/ft/fc/pattern.rs

Regarding the `emoji` font family, not sure if `fontconfig` supports "building" a fallback with a pattern, for example this matches, but the order is not right:

```
fc-match -s "monospace,emoji:style=Regular"
```

From reading around and going through my config, the `emoji` font family is supposed to be appended to other font families, but the monochrome "Noto Emoji" font still shows up in the middle of the other fonts (no idea why yet, imo it should not be there), while "Noto Color Emoji" is at the end, because it has been appended. I tried to add "Noto Emoji" to the preferred `emoji` fonts, so it is added at the very end as well, but without success.

In conclusion, how the `emoji` font family is used seem not to be a problem with Alacritty, but with `fontconfig` and/or the attributes of the Noto Emoji fonts. And on a separate issue, the `monospace` pattern may be incorrect

**Comment by maxbrunet** [comment]
Nevermind, I linked tests in my previous comment :see_no_evil:, and the Servo fontconfig libraries likely handle building patterns with the font family correctly.

I have run Alacritty with maximum verbosity (`alacritty -vvv`) and pasted an emoji in it, it appears to load `DejaVuSans` fonts at that point which must contain emojis.

So again a problem with the user's font config (mine). Anyway an explicit fallback to the Emoji font bumps its priority, solving the problem
